### ISSUE TITLE
Do negative samples participate in cls_loss calculation?

### ISSUE THREAD

**Issue by jshh0401** [ultralytics/yolov5#13580]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have fully trained a YOLOv5 model on a private dataset. When dealing with the false positives, I found that `x[:, 5:] *= x[:, 4:5]  # conf = obj_conf * cls_conf` . Then I visualized the cls_conf of negative samples and found that for a large number of negative samples, cls_conf is close to 1. I found that since negative samples do not participate in the calculation of cls loss, the model tends to output the probability of 1. Intuitively, I think cls_conf -> 1 / num_classes would be better. Is this correct?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jshh0401, thank you for your interest in YOLOv5 🚀! Please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/).

If this is a 🐛 Bug Report, please provide a **minimum reproducible example** to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

## Requirements

[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

🔔 This is an automated response to help guide your issue. An Ultralytics engineer will review and assist you soon!

**Comment by pderrenger** [comment]
You're making an astute observation about the classification confidence behavior in YOLOv5. You're correct that negative samples (background) don't participate in the cls_loss calculation, which can lead to the model outputting arbitrary classification confidences for negative samples - sometimes with values close to 1 for certain classes.

Your intuition that cls_conf should ideally tend toward 1/num_classes for negative samples is reasonable from a probabilistic perspective. This would indicate maximum uncertainty across all classes when no object is present. However, YOLOv5 relies on the objectness score (obj_conf) to filter out these false positives, rather than enforcing uniform classification distribution on negatives.

If you're dealing with many false positives, you might:

1. Try increasing your confidence threshold during inference
2. Adjust the balance between positive/negative samples during training with the `--hyp obj` hyperparameter
3. Consider using more augmentation to improve generalization

The final confidence calculation (obj_conf * cls_conf) is designed to help mitigate this issue by downweighting detections with low objectness, even if they have high class confidence.

**Comment by jshh0401** [comment]
Thank you for your answer and suggestions!

I understand that YOLOv5 uses obj_conf to filter negative samples. In reality, we hope to perform detection on low-computing and low-resolution platforms. The fewer FPs, the better, and at the same time, we need to ensure recall (although this is difficult to do). Currently, FP < 5%.

Before this, data enhancement has been done. Increasing the threshold will reduce recall (now threshold = 0.25). Adjusting the proportion of positive and negative samples can reduce FP and also reduce recall. Most methods of reducing FP will also reduce recall, a trade-off.

I want to know why we only rely on obj_conf to filter FP without considering negative samples.

Will negative samples participate in loss calculations and greatly affect the learning of positive samples when controlling cls_neg_loss <= cls_pos_loss?

Add negative samples to cls loss. If there are too many negative samples involved in the calculation, the positive samples may be overwhelmed. Therefore, I plan to take the Top K to ensure that cls_neg_loss<=cls_pos_loss.

Thanks for your great work and looking forward to your suggestions!

Best regards.

**Comment by pderrenger** [comment]
Hi @jshh0401,

You've raised an excellent point about the trade-off between reducing false positives and maintaining recall in detection systems. YOLOv5 primarily uses obj_conf for filtering negatives because it's an efficient design choice that simplifies training and keeps the model focused on the primary task of identifying positive objects.

Regarding your approach to include negative samples in cls_loss: this is actually a reasonable modification to explore. By using a Top-K strategy where cls_neg_loss ≤ cls_pos_loss, you could potentially teach the model to be more discriminative without overwhelming the positive sample learning. 

This approach has been explored in some other detection frameworks and could help address the issue of high cls_conf values on negatives. Since you're targeting low-compute platforms where precision is critical, this modification might be worth implementing as an experiment.

If you try this modification, I'd be interested to hear your results, especially any improvements in precision while maintaining acceptable recall rates.

**Comment by jshh0401** [comment]
The tentative conclusion is that simply letting negative samples participate in the classification loss is not very useful, and negative examples are still classified with high confidence.

**Comment by jshh0401** [comment]
Another experimental conclusion is that multi-scale resolution training during training, such as increasing the input image resolution but keeping it unchanged during inference, can reduce FP and improve recall.

**Comment by pderrenger** [comment]
Thanks for sharing these experimental results! Your finding about negative samples still producing high classification confidence even when included in cls_loss is interesting and aligns with the inherent difficulty of teaching models what "not an object" looks like across all possible classes.

Your second observation about multi-scale training improving both FP reduction and recall is excellent - this works because the model learns more robust features across different scales during training, making it better at distinguishing true positives from false positives at inference time. This is a proven technique that's often underutilized for improving precision/recall trade-offs.

For your low-compute deployment scenario, you might also consider experimenting with different anchor configurations or using knowledge distillation from a larger model trained at higher resolution.
